### ISSUE TITLE
Generate enum or string literal union types from table.

### ISSUE THREAD

**Issue by zimme** [kristiandupont/kanel#275]
Postgraphile have this feature where it allows you to generate GraphQL enum types from each row in a specified table in the database schema.

Something similar for kanel would be nice. I haven't seen that this is possible, but maybe I missed it?

**Comment by kristiandupont** [comment]
No, Kanel only knows about the schema, not actual data in the database. Is this for when you create a table that basically works like a glorified enum? I could see the value in that..

**Comment by zimme** [comment]
Yeah, that's exactly it.
```
create table offers_status (
  name text primary key,
  description text
);

insert into offers_status (name, description) values ('ACCEPTED', 'Offer accepted'), ('DECLINED', 'Offer declined');
```
It would be nice if this could generate one or both of the following.
```
type OffersStatus = "ACCEPTED" | "DECLINED";

enum OffersStatus {
  ACCEPTED = "ACCEPTED",
  DECLINED = "DECLINED",
}
```

**Comment by kristiandupont** [comment]
I am curious to know if you have any way of unambiguously identifying these "enum tables"?
My intuition would be to use a tag in the comment string or perhaps a naming convention. Do you have anything?

**Comment by zimme** [comment]
Postgraphile are using what they are calling "Smart comments" which is basically @ prefixed comments in the database or a smart tags file.

In the case of enum tables the smart comment have to be in the database. `@enum` is required and `@enumName` and `@enumDescription` are optional.

e.g.
```
COMMENT ON TYPE animal_type IS E'@enum\n@enumName TypeOfAnimal';
```

And here's a link to the docs for enums in Postgraphile.
https://www.graphile.org/postgraphile/enums/

**Comment by kristiandupont** [comment]
Awesome, thank you!

**Comment by zimme** [comment]
Thinking about it some more from Kanel's perspective. If you don't want to rely on something like smart comments or similar. I would be perfectly happy to just have to add the "enum tables" in the kanel config file for kanel to be able to figure out which tables to generate enums or string literal types for.

**Comment by kristiandupont** [comment]
I am thinking that it will be an extension that you need to configure some function that tells you whether a table represents an enum or not.

The question that I am unsure of then, is what to do with the tables that "use" this enum. Would you just replace the reference with the enum? Or have both?

**Comment by kristiandupont** [comment]
Got it. I think I can make something that is sufficiently configurable there..

**Comment by zimme** [comment]
If you have time, feel free to write a rough outline of how one could implement this as a plugin. Perhaps with that I'm able to give it a try myself =)

**Comment by kristiandupont** [comment]
Hi Simon,

sorry about coasting on this! I am also slacking on other aspects of the project as you might have noticed. 

Let's see. You would probably want to make a Let's see. You will probably want to make a [preRenderHook](https://kristiandupont.github.io/kanel/preRenderHooks.html). 
I am happy to try and help, but here is at least a starting point:

```typescript

const applyEnumTables: PreRenderHook = async (outputAcc, instantiatedConfig) => {
  const connection = instantiatedConfig.connection;
  const db = knex({ client: 'postgres', connection });
  
  // Get all the tables in all the schemas  
  const allTables = Object.values(instantiatedConfig.schemas).map(({ table }) => table).flat();
  
  // ..and find the ones that are enums, however you define that.
  const enumTables = allTables.filter(table => table.comment.indexOf('@enum') !== -1);
  
  const overrides = {};
  
  for (let table of enumTables) {
    const columns = await db.select('*').from(table.name);
    
    const declaration: GenericDeclataion = {
      declarationType: 'generic',
      lines: [
        `enum ${table.name} {`,
        ...columns.map(col => `  ${col.name}`),
        `};`,
        '',
        `export default ${table.name};`
      ]
    };
    const path = join(instantiatedConfig.outputPath, table.name);
    
    overrides[path] = { declarations: [declaration] };
  }
  
  
  db.destroy();

  return {
    ...outputAcc,
    ...overrides
  };
}
```

This is UNTESTED, even uncompiled code that I just wrote up here, so it will not work. But it might get you something to go by.
Basically, it should generate a file for every table that has a comment that includes an enum tag. Still missing is actually referencing the generated types everywhere, I will need to look into that a bit later :-)

**Comment by zimme** [comment]
That's completely fine. Unless your day job is working on open source projects it's kinda expected that people work on things sporadically.

I'll try and digg into this, this weekend hopefully.

**Comment by zimme** [comment]
This seems to work and should work with different types of enumStyle.

Ignore the type comments as I tested this in the kanelrc.js file :D

This seems to do what I was thinking... However... Perhaps, there should be some combination of this and the normal table types generated by kanel that would be the "best" way to generate types for tables with an "@enum" comment.

That way, you would get types for interacting with the table as well as enum types and it would be nice if those could even be used in any related table so instead of "string" those would be typed to use the enum type. I can probably do a better job of explaining what I'm thinking.

```
async (outputAccumulator, instantiatedConfig) => {
  const connection = instantiatedConfig.connection;
  const db = knex({ client: "postgres", connection });

  // Get all the tables in all the schemas
  const allTables = Object.values(instantiatedConfig.schemas)
    .map(({ tables }) => tables)
    .flat();

  // ..and find the ones that are enums, however you define that.
  const enumTables = allTables.filter(
    (table) => table.comment !== null && table.comment.indexOf("@enum") !== -1
  );

  /** @type {import("kanel").Output} */
  const overrides = {};

  for (let table of enumTables) {
    const { comment, name, path } = instantiatedConfig.getMetadata(
      table,
      "enum",
      instantiatedConfig
    );

    const rows = await db
      .withSchema(table.schemaName)
      .select("name")
      .from(table.name);

    if (instantiatedConfig.enumStyle === "type") {
      /** @type {import("kanel").TypeDeclaration} */
      const declaration = {
        declarationType: "typeDeclaration",
        name,
        comment,
        exportAs: "default",
        typeDefinition: [
          "", // Start definition on new line
          ...rows.map((row) => `| '${row.name}'`),
        ],
      };

      overrides[path] = { declarations: [declaration] };
    } else {
      /** @type {import("kanel").GenericDeclaration} */
      const declaration = {
        declarationType: "generic",
        comment,
        lines: [
          `enum ${name} {`,
          ...rows.map((row) => `  ${escapeName(row.name)} = '${row.name}',`),
          "};",
          "",
          `export default ${name};`,
        ],
      };

      overrides[path] = { declarations: [declaration] };
    }
  }

  db.destroy();

  return {
    ...outputAccumulator,
    ...overrides,
  };
}
```

**Comment by zimme** [comment]
So the normal type for a "enum" table named features would be this.
```
// @generated
// This file is automatically generated by Kanel. Do not modify manually.
/** Identifier type for app_public.features */
export type FeatureName = string &  { __flavor?: 'FeatureName' };
/** Represents the table app_public.features */
export default interface Feature {
  name: FeatureName;
  description: string | null;
}
/** Represents the initializer for the table app_public.features */
export interface FeatureInitializer {
  name: FeatureName;
  description?: string | null;
}
/** Represents the mutator for the table app_public.features */
export interface FeatureMutator {
  name?: FeatureName;
  description?: string | null;
}
```

Perhaps it's only the `type FeatureName = string & { __flavour?: 'FeatureName' };` part that should be changed to be
```
type FeatureName =
  | 'WAITLIST'
  | 'WISHLIST';
```
or something similar...

also, if enumStyle isn't `type` I guess we should be generating enum instead of type.

**Comment by zimme** [comment]
I guess the mutator and Initializer types would need to be `string` instead of `FeatureName` otherwise you would not be able to add new values to the table as the only valid types are only the rows that already exists :P

**Comment by kristiandupont** [comment]
Nice! 

I guess you could just replace the identifier type. I guess I would suggest you make it work in your own setup first. Then we can see if we can generalize this enough to be a plugin. If not, it would make a good blog post :-D

If you do end up making a plugin, I would be happy to either include it in this repository or link to your repository, whichever you prefer.

**Comment by zimme** [comment]
So I can quite easily find and override the type for the identifier/primary key column type. However It seems to get messier when I want to override the Initializer and Mutator types, to use "string" instead of "FeatureName" (from the previous example)

Any pointers how to find the interface definition for the initializer/mutator and override it, without it being super brittle depending on what people might do in getMetadata/getPropertyMetadata?

**Comment by zimme** [comment]
Saw something in kanel-knex, that might be helpful here... I'll see if that helps me in some way.

**Comment by kristiandupont** [comment]
Right, so I don't know how clear it is but the thing that you return from your hook contains all the generated files with all of their contents. So you can either override a single file or some of the contents of a file. The interfaces for the tables consist of "interface declarations". These have properties that in turn may or may not have type imports. So what you will probably want to do is find all the affected tables, replace the property that references the enum with whatever you created, including the new type import.

This might be a bit hard to understand, I know it's a bit of an unusual pattern. I am happy to elaborate (or jump on a quick call or something), if you want.

**Comment by zimme** [comment]
Initial version available here #350.

**Comment by zimme** [comment]
I'll add a reminder to rebase the PR on main and see if any changes are needed to get it finished.
